crates/arrow_tools/src/schema.rs

```diff
@@ -160,6 +160,71 @@ pub fn schema_meta_get_computed_columns(
     }
 }
 
+/// Returns a string describing the difference between two schemas.
+#[must_use]
+pub fn schema_difference(expected: &Schema, actual: &Schema) -> String {
+    let mut differences = Vec::new();
+
+    // Check for missing columns in actual schema
+    for field in expected.fields() {
+        if !actual.fields().iter().any(|f| f.name() == field.name()) {
+            differences.push(format!("The column `{}` is missing", field.name()));
+        }
+    }
+
+    // Check for extra columns in actual schema
+    for field in actual.fields() {
+        if !expected.fields().iter().any(|f| f.name() == field.name()) {
+            differences.push(format!("The column `{}` is unexpected", field.name()));
+        }
+    }
+
+    // Check for type mismatches in common columns
+    for expected_field in expected.fields() {
+        if let Some(actual_field) = actual
+            .fields()
+            .iter()
+            .find(|f| f.name() == expected_field.name())
+        {
+            if !DFSchema::datatype_is_semantically_equal(
+                expected_field.data_type(),
+                actual_field.data_type(),
+            ) {
+                differences.push(format!(
+                    "The type of `{}` changed from `{}` to `{}`",
+                    expected_field.name(),
+                    expected_field.data_type(),
+                    actual_field.data_type()
+                ));
+            }
+
+            // Check for nullability changes
+            if expected_field.is_nullable() != actual_field.is_nullable() {
+                differences.push(format!(
+                    "The column `{}` changed from {} to {}",
+                    expected_field.name(),
+                    if expected_field.is_nullable() {
+                        "nullable"
+                    } else {
+                        "non-nullable"
+                    },
+                    if actual_field.is_nullable() {
+                        "nullable"
+                    } else {
+                        "non-nullable"
+                    }
+                ));
```

Comment:
```suggestion
                differences.push(format!(
                    "The column `{column_name}` changed from {expected_nullable} to {actual_nullable}",
                    column_name = expected_field.name(),
                    expected_nullable = if expected_field.is_nullable() {
                        "nullable"
                    } else {
                        "non-nullable"
                    },
                    actual_nullable = if actual_field.is_nullable() {
                        "nullable"
                    } else {
                        "non-nullable"
                    }
                ));
```

some of these `format!`'s are complex. Can we use named parameters?